### PR TITLE
Matrix matrix multiplication layer.

### PR DESCRIPTION
**Pull request by dominikgrewe**
Multiplies two matrices (or two batches of matrices).
This is a pure Lua implementation taking advantage of the new torch.bmm
function for minibatch inputs.

### CODE REVIEW

**Comment by soumith** [comment]
this seems slightly short-sighted. 
Would it not be much better if there's one module called MatrixMatrixMultiply, or nn.mm that either took in a single matrix-pair os a batch of matrix-pairs (just like many other nn modules which take a single sample or a batch of samples....)

will let others comment as well.

**Comment by dominikgrewe** [comment]
Fair point. If the policy is that every module should support single inputs, then we should do that here too.
I've updated the PR.

File: MatrixMatrixMultiply.lua [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  self.gradInput = {torch.Tensor(), torch.Tensor()}
+end
+
+function MatrixMatrixMultiply:updateOutput(input)
+  assert(#input == 2, 'input must be a pair of minibatch matrices')
+  local a, b = unpack(input)
+  assert(a:nDimension() == 2 or a:nDimension() == 3, 'arguments must be a 2D or 3D Tensor')
+
+  if a:nDimension() == 2 then
+    assert(b:nDimension() == 2, 'arg 2 must be a 3D Tensor')
```
Comment (nicholas-leonard) [inline MatrixMatrixMultiply.lua]: 3D Tensor?

Comment (soumith) [inline MatrixMatrixMultiply.lua]: Yes 3D for batch mm

Comment (soumith) [inline MatrixMatrixMultiply.lua]: Ah I see for non-batch it is printed wrong

**Comment by dominikgrewe** [comment]
Good catch, thanks! I've updated that one and also below.

**Comment by soumith** [comment]
The last comment from me is about the naming.

Traditionally, the nn package has nn.dot, or nn.CAdd or nn.Add, nn.Select etc, which imply that it is doing what it's torch equivalent would do.
If that spirit is maintained, this should be named nn.mm , what do you guys think...

**Comment by soumith** [comment]
or nn.MM

**Comment by nicholas-leonard** [comment]
nn.MM

**Comment by dominikgrewe** [comment]
Okay, I've renamed the module "nn.MM"

File: MM.lua [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  assert(#input == 2, 'input must be a pair of minibatch matrices')
+  local a, b = unpack(input)
+  self.gradInput[1]:resizeAs(a)
+  self.gradInput[2]:resizeAs(b)
+
+  assert(gradOutput:nDimension() == 2 or gradOutput:nDimension() == 3, 'arguments must be a 2D or 3D Tensor')
+
+  local h_dim, w_dim, f
+  if gradOutput:nDimension() == 2 then
+    assert(a:nDimension() == 2, 'arg 2 must be a 2D Tensor')
```
Comment (nicholas-leonard) [inline MM.lua]: Could you make the messages here more specific? Like arg 2a vs arg 2b or something.

File: MM.lua [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    if self.transA then a = a:t() end
+    if self.transB then b = b:t() end
+    assert(a:size(2) == b:size(1), 'matrix sizes do not match')
+
+    self.output:resize(a:size(1), b:size(2))
+    self.output:mm(a, b)
+  else
+    assert(b:nDimension() == 3, 'arg 2 must be a 3D Tensor')
+    assert(a:size(1), b:size(2), 'inputs must contain the same number of minibatches')
```
Comment (nicholas-leonard) [inline MM.lua]: `a:size(1) == b:size(1)`?

File: doc/simple.md [hunk 4]
```diff
@@ -831,3 +832,14 @@ gnuplot.grid(true)
...
+## MM ##
+
+`module` = `nn.MM(transA, transB)`
+
+Performs multiplications on one or more pairs of matrices.
+If `transA` is set, the first matrix is transposed before multiplication.
+If `transB` is set, the second matrix is transposed before multiplication.
+By default, the matrices do not get transposed.
+
+The module also accepts 3D inputs which are interpreted as batches of matrices.
```
Comment (nicholas-leonard) [inline doc/simple.md]: Can you mention what axis order is expected :  A : b x m x n, B: b x n x m. It makes it easier to understand. Maybe add a small example.

**Comment by nicholas-leonard** [comment]
Very cool module.

File: MM.lua (hunk 2 above)
Comment (dominikgrewe) [inline MM.lua]: I've change the message to hopefully make it clearer.

File: MM.lua (hunk 3 above)
Comment (dominikgrewe) [inline MM.lua]: Good catch, thanks!

File: doc/simple.md (hunk 4 above)
Comment (dominikgrewe) [inline doc/simple.md]: I've extended the documentation a bit and also provided an example.

**Comment by soumith** [comment]
@dominikgrewe looks good to me. squash and will merge.

**Comment by dominikgrewe** [comment]
Thanks! Squashed.

**Comment by soumith** [comment]
thanks, great PR!

**Pull request merged**